src/PooledArrays.jl

```diff
@@ -341,16 +417,19 @@ Base.@propagate_inbounds function Base.isassigned(pa::PooledArray, I::Int...)
     !iszero(pa.refs[I...])
 end
 
-# Vector case
-Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real,AbstractVector}...)
-    PooledArray(RefArray(getindex(A.refs, I...)), copy(A.invpool))
+# Other cases; we rely on the fact that non-standard indexing will fall back to this eventually
+Base.@propagate_inbounds function Base.getindex(A::PooledArray, I::Union{Real, AbstractVector}...)
+    # make sure we do not increase A.refcount in case creation of newrefs fails
+    newrefs = getindex(A.refs, I...)
+    @assert newrefs isa AbstractArray
+    Threads.atomic_add!(A.refcount, 1)
+    return PooledArray(RefArray(newrefs), A.invpool, A.pool, A.refcount)
 end
 
-# Dispatch our implementation for these cases instead of Base
-Base.@propagate_inbounds Base.getindex(A::PooledArray, I::AbstractVector) =
-    PooledArray(RefArray(getindex(A.refs, I)), copy(A.invpool))
-Base.@propagate_inbounds Base.getindex(A::PooledArray, I::AbstractArray) =
-    PooledArray(RefArray(getindex(A.refs, I)), copy(A.invpool))
+# views
+
+Base.view(A::PooledArray, I...) =
```

Comment:
@nalimilan - before moving forward this is another change I propose for performance reasons. This is for sure breaking (fortunately mildly), so I bumped version to 2.0.0.

The point is that in this way we always can correctly detect if something is a `PooledArray` even if someone takes its view.

I think similar thing can be done in CategoricalArrays.jl and then we totally can avoid `CatArrOrSub` problems. I also think this is something that MLJ.jl would prefer. CC @ablaom

Reply:
This implementation is overly simplified, so I would have to fix it (which is doable).

Before I do it let us decide which we prefer:
* `view` of `PooledArray` is `PooledArray` (and nothing has to be done except making sure the implementation is correct)
* `view` of `PooledArray` is `SubArray` (and then we have to add special methods for `getindex`, `copyto!` etc. for views of `PooledArray`)

Reply:
Can leave this for another PR? I'm not sure what's the expected API for `view`, but `?view` and `?@view` seem to assume that a `SubArray` is returned.

Reply:
I need it in this PR, but as discussed on Slack I will change the implementation to `SubAarray`